### PR TITLE
Nullable annotations refactoring helpers & bits and pieces elsewhere

### PR DESCRIPTION
**Pull request by petrroll**
Based on #38166

### CODE REVIEW

**Comment by petrroll** [comment]
New commits from: f65a165

File: src/Features/CSharp/Portable/CodeRefactorings/CSharpRefactoringHelpersService.cs [hunk 1]
```diff
@@ -7,13 +7,19 @@
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.LanguageServices;
 
+#nullable enable
```
Comment (jasonmalinowski) [inline src/Features/CSharp/Portable/CodeRefactorings/CSharpRefactoringHelpersService.cs]: I think we've been putting these before the using block. Even if not that, blank line after.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs [hunk 2]
```diff
@@ -262,7 +264,9 @@ internal interface ISyntaxFactsService : ILanguageService
         bool IsDirective(SyntaxNode node);
         bool IsForEachStatement(SyntaxNode node);
         bool IsLockStatement(SyntaxNode node);
-        bool IsUsingStatement(SyntaxNode node);
+#nullable enable
+        bool IsUsingStatement(SyntaxNode? node);
+#nullable disable
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs]: These should all be #nullable restore -- when we flip it on for the project globally we don't want these staying disabled.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs [hunk 3]
```diff
@@ -33,6 +34,7 @@ internal interface ISyntaxFactsService : ILanguageService
         bool IsPredefinedType(SyntaxToken token, PredefinedType type);
         bool IsPredefinedOperator(SyntaxToken token);
         bool IsPredefinedOperator(SyntaxToken token, PredefinedOperator op);
+#nullable enable
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs]: Already was enabled?

File: src/Features/CSharp/Portable/CodeRefactorings/CSharpRefactoringHelpersService.cs (hunk 1 above)
Comment (petrroll) [reply src/Features/CSharp/Portable/CodeRefactorings/CSharpRefactoringHelpersService.cs]: Yep, checked how it's done elsewhere, fixed.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs (hunk 3 above)
Comment (petrroll) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs]: Yep, that's a bug there.

File: src/Features/Core/Portable/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs [hunk 4]
```diff
@@ -51,6 +50,7 @@ public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContex
                 return;
             }
 
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
             context.RegisterRefactoring(
                 new MyCodeAction(GetTitle(),
                     c => InvertIfAsync(root, document, ifNode, c)),
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs]: note: i don't know why we're passing the root in here.  could just be removed.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs [hunk 5]
```diff
@@ -551,6 +551,7 @@ protected int GetStartOfNodeExcludingAttributes(SyntaxNode node)
 
         public abstract SyntaxList<SyntaxNode> GetAttributeLists(SyntaxNode node);
 
+#nullable enable
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs]: i'd personally prefer a whole-file approach to nullable.  If we're not ready for the full file, then don't bother.  once we are, we can just do the whole file and move on.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs [hunk 6]
```diff
@@ -24,6 +24,7 @@ internal interface ISyntaxFactsService : ILanguageService
 
         SyntaxToken ParseToken(string text);
 
+#nullable enable
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs]: same here.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs [hunk 7]
```diff
@@ -262,7 +264,9 @@ internal interface ISyntaxFactsService : ILanguageService
         bool IsDirective(SyntaxNode node);
         bool IsForEachStatement(SyntaxNode node);
         bool IsLockStatement(SyntaxNode node);
-        bool IsUsingStatement(SyntaxNode node);
+#nullable enable
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs]: this is just getting messy and unpleasant IMO.  just come and do this full file at some point the future

File: src/Features/Core/Portable/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs (hunk 4 above)
Comment (petrroll) [reply src/Features/Core/Portable/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs]: That's a previous PR change, will take a look at it.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs (hunk 5 above)
Comment (petrroll) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs]: I see your point but I think there's a benefit in creating islands of nullability whenever some code is touched. 

That said, I don't have strong preference on this one, just following an "ok" to move even "parts of file" by @333fred

Comment (333fred) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs]: At least in the compiler, we are perfectly fine with enabling nullability in small sections as we work on related pieces of code.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs]: I took a look at these files, and i personally don't get the value of the island.  They're not uber complex areas with lots of moving parts.  We could just turn it on for the file, do the few minutes or work to make it clean, and just know we were done with it.

Comment (petrroll) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs]: TBH, not knowing the usage of most of SyntaxFacts API it's not that quick for me to decide whether we want / can have this parameter nullable or not.

I'm happy to remove the nullability island and keep just my whole files #nullable enabled if that's what's desirable.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs]: i'd prefer that TBH.  We can bit off syntaxfacts in one go in the future.  I'd be happy to help out there myself.  

Note: the semantics *are* supposed to be that effectively it is 'null tolerant' on all query methods (i.e. IsFoo).  But that it is null-interolerant on grabbing data off a node.  so "GetExpressionOfExpressionStatemnet" expects that you already validated what you passed in was an ExpressionStatement.

Comment (petrroll) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs]: Ok, will cleanup tomorrow.

**Comment by petrroll** [comment]
cc @jasonmalinowski fixed

File: src/Features/Core/Portable/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs [hunk 8]
```diff
@@ -50,6 +50,7 @@ public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContex
                 return;
             }
 
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
```
Comment (mavasani) [inline src/Features/Core/Portable/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs]: Unnecessary?

**Pull request merged**